### PR TITLE
Limit the size and age of the ZIP-401 rejected transaction ID list

### LINKED ISSUE #2759: Limit the size and age of the ZIP-401 rejected transaction ID list
### Motivation

ZIP-401 limits the number of transactions in the rejected list, and the age of the entries in that list.

### This Ticket Depends On

`RecentlyEvicted`:
  - [x] #2780
    - [x] #2778
    - [x] #2779
      - [x] #2440

### Specification

The checked items are covered by other tickets.

> `EvictTransaction` MUST do the following:
> - [x] Select a random transaction to evict, with probability in direct proportion to `eviction weight`.
> - [x] Add the `txid` and the current `time` to `RecentlyEvicted`,
>   - [ ] dropping the oldest entry in `RecentlyEvicted` if necessary to keep it to at most `eviction_memory_entries` entries.
> - [x] Remove [the randomly selected transaction] from the mempool.

> - [ ] Nodes SHOULD remove transactions from `RecentlyEvicted` that were evicted more than `mempoolevictionmemoryminutes` minutes ago. This MAY be done periodically, and/or just before `RecentlyEvicted` is accessed when receiving a transaction.

https://zips.z.cash/zip-0401#specification

### Design

The ZIP-401 reject list must be implemented as a new list in `MempoolStorage`. This avoids bugs that confuse different rejection rules.

Since this time is only used by the local node, it should be monotonic (strictly increasing):
https://docs.rs/tokio/1.11.0/tokio/time/struct.Instant.html

If we use an `IndexMap<transaction::Hash, Instant>`, we can efficiently evict transaction IDs by using `IndexMap::pop`, until we reach a transaction that's below the eviction limit: 
https://docs.rs/indexmap/1.7.0/indexmap/map/struct.IndexMap.html#order

### LINKED ISSUE #2958: Remove MEMPOOL_SIZE and just rely on the ZIP-401 cost limit
## Motivation

The current `MEMPOOL_SIZE` of 4 means that we'll never actually reach the ZIP-401 mempool cost limit.

## Tasks

- Remove `MEMPOOL_SIZE`
- Quote the cost limit rule from ZIP-401 where it's actually used in the code

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

ZIP-401 specifies that the list of TXIDs must have a maximum size and entries older than a specified threshold must be removed.

### Specifications

https://zips.z.cash/zip-0401#specification

### Designs

Some of the design was discussed in https://github.com/ZcashFoundation/zebra/issues/2759

## Solution

- Create a `EvictedList` that encapsulates both behaviours
- Use the `EvictedList` for both rejected and expired transactions
  - I'm using in both because that was easier with the current `Storage` design, and seems harmless (expired transactions will also have a timeout)
  
I tried creating a generic `OrderedMap` data structure, that `EvictedList` would use, in order to organize the code. However I noticed it would be more work since it's tricky to handle/document the behavior that removing a item from the list might not _really_ remove a item from the list (when a value was refreshed and an stale entry is left behind). So I left the entire logic inside `EvictedList` (where the removal methods are not exposed).

Closes https://github.com/ZcashFoundation/zebra/issues/2759
Closes #2958

## Review

Anyone can review. @teor2345 discussed the design.

This may require adjustments after https://github.com/ZcashFoundation/zebra/pull/2889/ so it may be better to wait for that to be merged first.

### Reviewer Checklist

  - [ ] Does it make sense to use `EvictList` for the expired transaction, or should change `Storage` so that only rejected list uses it?
  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebrad/src/components/mempool/storage.rs [hunk 1]
```diff
@@ -125,8 +126,11 @@ pub struct Storage {
     /// These rejections apply until a rollback or network upgrade.
     ///
     /// Any transaction with the same `transaction::Hash` is invalid.
-    chain_rejected_same_effects:
-        HashMap<SameEffectsChainRejectionError, HashSet<transaction::Hash>>,
+    chain_rejected_same_effects: HashMap<SameEffectsChainRejectionError, EvictionList>,
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:133]: Design Change or Docs: We don't need to use an `EvictionList` for `Expired` transactions, because expiry is permanent until a chain fork or network upgrade. And we handle those by clearing the mempool.

But it also doesn't matter if `Expired` rejections only persist for 60 minutes, because that's a long time, and validation of expiry is done early and cheaply.

I don't mind what we do here.
But can we document this behaviour somewhere?

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    collections::{HashMap, VecDeque},
+    time::{Duration, Instant},
+};
+
+use zebra_chain::transaction;
+
+/// An eviction list that allows to efficiently add entries, get entries,
+/// and remove older entries in the order they were inserted.
+pub struct EvictionList {
+    // Maps each TXID In the list to the instant they were added.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Docs: clarification, typo

```suggestion
    // Maps each TXID in the list to the most recent instant they were added.
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    unique_entries: HashMap<transaction::Hash, Instant>,
+    // The entries in the order they were inserted.
+    // This can be larger than `unique_entries` if a same txid is added
+    // multiple times. Its instant will be overwritten in
+    // `unique_entries` but all entries will kept in `unique_entries`.
+    ordered_entries: VecDeque<(transaction::Hash, Instant)>,
+    // The maximum size of `unique_entries`.
+    size: usize,
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Nitpick: We usually use `size` or `len` for the current size, and `max_size` for the limit.

```suggestion
    max_size: usize,
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    /// The mempool transaction eviction age limit.
+    /// Same as [`Config::eviction_memory_time`].
+    eviction_time: Duration,
+}
+
+impl EvictionList {
+    /// Create a new [`EvictionList`] with the given maximum size and
+    /// eviction time.
+    pub fn new(size: usize, eviction_time: Duration) -> Self {
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Nitpick: naming

```suggestion
    pub fn new(max_size: usize, eviction_time: Duration) -> Self {
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    ///
+    /// All entries older than [`EvictionList::eviction_time`] will be removed.
+    pub fn insert(&mut self, key: transaction::Hash) {
+        self.prune_old();
+        while self.len() >= self.size {
+            self.pop_front();
+        }
+        let value = Instant::now();
+        let old_value = self.unique_entries.insert(key, value);
+        if old_value != Some(value) {
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Nice catch!

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            self.ordered_entries.push_back((key, value))
+        }
+    }
+
+    /// Checks if the given TXID is in the list.
+    pub fn contains_key(&self, txid: &transaction::Hash) -> bool {
+        self.unique_entries.contains_key(txid)
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Likely bug: Do we need to prune old entries here, or skip entries that have expired?

Since we don't, I think we can call `contains_key`, and get expired entries.

Can you also add a test for this bug?
(Or remove `prune_old()` from the existing tests?)

Alternative impls:

1. Just check for expiry but don't prune:

```rust
        let now = Instant::now();
        if let Some(evicted_at) = self.unique_entries.get(txid) {
            if (now - *evicted_at) <= self.eviction_time {
                return true;
            }
        }

        false
```

We might also want to turn `(now - *evicted_at) <= self.eviction_time` into an `EvictionList` method, because it happens here and in `prune_old`.

2. Needs `&mut self`:

```rust
        self.prune_old();
        self.unique_entries.contains_key(txid)
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /// Clear the list.
+    pub fn clear(&mut self) {
+        self.unique_entries.clear();
+        self.ordered_entries.clear();
+    }
+
+    /// Prune TXIDs that are older than `eviction_time` ago.
+    pub fn prune_old(&mut self) {
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs:103]: Visibility Nitpick: This seems like an internal method?

```suggestion
    pub(super) fn prune_old(&mut self) {
```

**Review by teor2345 — COMMENTED** [review]
This looks good, but I think `contains_key` can return transaction IDs that have expired since the last `insert`.

I also have a few minor suggestions, feel free to ignore them.

File: zebrad/src/components/mempool/storage.rs (hunk 1 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs:133]: Done in ecfd0a3c

File: zebrad/src/components/mempool/storage/eviction_list.rs (hunk 6 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: Great catch!

I went with the first approach in ecfd0a3c. It seemed unfortunate to make the method mutable.

I also fixed `len()` which had the same issue.

File: zebrad/src/components/mempool/storage/eviction_list.rs (hunk 7 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs:103]: My rationale is that ZIP-401 states 

> (...) This MAY be done periodically

So if we decide to do it periodically for whatever reason, we can use the method. Added a comment about that in ecfd0a3c

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 8]
```diff
@@ -0,0 +1,138 @@
+//! [`EvictionList`] represents the transaction eviction list with
+//! efficient operations.
+use std::{
+    collections::{HashMap, VecDeque},
+    time::{Duration, Instant},
+};
+
+use zebra_chain::transaction;
+
+/// An eviction list that allows to efficiently add entries, get entries,
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Nitpick: Typo?

```suggestion
/// An eviction list that allows Zebra to efficiently add entries, get entries,
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+/// and remove older entries in the order they were inserted.
+pub struct EvictionList {
+    // Maps each TXID in the list to the most recent instant they were added.
+    unique_entries: HashMap<transaction::Hash, Instant>,
+    // The entries in the order they were inserted.
+    // This can be larger than `unique_entries` if a same txid is added
+    // multiple times. Its instant will be overwritten in
+    // `unique_entries` but all entries will kept in `unique_entries`.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Incorrect documentation / typo: All entries are kept in `ordered_entries`.

```suggestion
    // `unique_entries` but all entries will kept in `ordered_entries`.
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            max_size,
+            eviction_memory_time,
+        }
+    }
+
+    /// Inserts a TXID in the list, keeping track of the time it was inserted.
+    ///
+    /// If the TXID is already in the list, its insertion time will be updated.
+    ///
+    /// All entries older than [`EvictionList::eviction_time`] will be removed.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Nitpick: missed docs in field rename?

```suggestion
    /// All entries older than [`EvictionList::eviction_memory_time`] will be removed.
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        // > Nodes SHOULD remove transactions from RecentlyEvicted that were evicted more than
+        // > mempoolevictionmemoryminutes minutes ago. This MAY be done periodically,
+        // > and/or just before RecentlyEvicted is accessed when receiving a transaction.
+        self.prune_old();
+        // > Add the txid and the current time to RecentlyEvicted, dropping the oldest entry
+        // > in RecentlyEvicted if necessary to keep it to at most eviction_memory_entries entries.
+        // Use a `while` because it is possible that more than one item is removed
+        // (if an entry is refreshed, it leaves the old value in the list).
+        while self.len() >= self.max_size {
+            self.pop_front();
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Possible CPU DoS bug: `len()` is `O(len)` here.

Can we make `len()` into an `O(expired)` operation using `prune_old()` and `unique_entries.len()`?

(And since we've called `prune_old()` already here, `len()` would be `O(1)`.)

<details>

`len()` is now `O(len + expired)`, so `insert()` is `O((len + expired) * (expired + 1)))`.

Which means `insert()` is `O(n^2)` in the worst case scenario. But `expired` is zero here, because we just called `prune_old()`. Even under these conditions, it's still `O(len)`.

If we don't fix the complexity of `len()`, we'd need to document the `O(len)` complexity of `insert()`.

But `insert()` is already called in a loop when we get new verified transactions. So that's `O(len * new_transactions)` complexity. Which we need to avoid to meet our DoS goals.

</details>

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        }
+        let value = Instant::now();
+        let old_value = self.unique_entries.insert(key, value);
+        if old_value != Some(value) {
+            self.ordered_entries.push_back((key, value))
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Likely RAM DoS bug: `ordered_entries` can grow very large.

For example, peers could add `max_size` entries in each new `Instant`. If the set of added entries is unique within instants, but identical between them, then on a system with a `1 ms` clock resolution, `ordered_entries` can have `eviction_memory_time * 1000 * max_size` entries. (And most systems have sub-millisecond resolutions.)

I suggest that we drop the earliest entries in `ordered_entries` if it exceeds `max_size`. We should remove any matching `unique_entries` as well, so that `unique_entries` does not exceed `max_size`.

Can you also add a test that makes sure `ordered_entries` and `unique_entries` never exceed their maximum sizes?

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    /// Get the size of the list.
+    pub fn len(&self) -> usize {
+        // Since the list is pruned only in mutable functions, make sure
+        // we take expired items into account.
+        let expired = self
+            .unique_entries
+            .iter()
+            .take_while(|(_txid, evicted_at)| self.has_expired(*evicted_at))
+            .count();
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Likely logic bug: `HashMap::iter()` visits entries in arbitrary order:
https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.iter

Can we simplify this method using `prune_old()` and `unique_entries.len()`?

<details>

If we don't simplify it, we'd need to visit entries in expiry order, then check their latest eviction times for expiry. And write a bunch of tests for each edge case.

Even if we did all these things, this method would still be `O(n)`, and there's no easy way to test for unexpected O(n^2) behaviour. This is a big risk in DoS-prevention code.

```suggestion
        let expired = self
            .ordered_entries
            .iter()
            .filter_map(|(txid, _ordered_evicted_at)| self.unique_entries.get(&txid).map(|latest_evicted_at| (txid, unique_evicted_at)))
            .take_while(|(txid, latest_evicted_at)| self.has_expired(*latest_evicted_at))
            .count();
```

</details>

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+            // we take expired items into account.
+            if !self.has_expired(evicted_at) {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Get the size of the list.
+    pub fn len(&self) -> usize {
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Possible CPU DoS bug: `len()` is now `O(len + expired)`.

We already call `len()` inside a loop in `insert()`, which is inside another loop of verified transactions. This is potentially `O(n^3)` complexity.

Can we make `len()` into an `O(expired)` operation using `prune_old()`?

<details>

If we don't fix the method, we'd need to:
* document its complexity,  since it's the only method that's not `O(1)`
* make sure that `len()` is never called in a loop, unless `prune_old()` has been called first
* make it into a private method, to make it easier to make sure that `len()` is used correctly

Even if we did all these things, there's no easy way to test for unexpected `O(n^2)` behaviour. So we'd be risking DoS bugs every time we changed this code.

</details>

**Review by teor2345 — COMMENTED** [review]
Nice catch that we might need to remove multiple entries in `insert()`!

I think we might need to `prune_old()` in `len()` to meet our DoS performance goals.
(I didn't realise this until you fixed `len()`.)

I also think we need to limit the length of `ordered_entries`, even if its entries haven't expired.

File: zebrad/src/components/mempool/storage/eviction_list.rs (hunk 13 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: Oops :man_facepalming: 

Thanks for catching this. I meant to write a O(expired) method, that's why I used `take_while`.

By using your suggestion but calling `take_while` before filtering, I believe it can be done in O(expired). I implemented that 05beee87

However, thinking more carefully about making the method mutable, I noticed that it would only "infect" the storage (I feared I'd need to make a lot of functions mutable too all the way to the Mempool). The only downside is that `Storage::rejected_transaction_count` needs to be mutable, but it's not yet used elsewhere. For this reason I overwrote the above with your suggestion of just calling `prune_old()` in 1ea2d12. If you think the approach above is better, it's easy to revert the later one.

Just to make it clearer:

- 05beee87: `len()` is not mutable and computes the real size on-the-fly. Upside: does not need to be mutable. Downside: error-prone, less efficient (e.g. if there a lot of expired entries, and no items are added in the EvictedList for a while, each call will be `O(expired)`)
- 1ea2d12 (the current implementation in the branch): `len()` is mutable and just calls `prune_old()`. Upside: simpler, more efficient (in that scenario, the first call will be `O(expired)` and others will be `O(1)`). Downside: mutable, forces `Storage::rejected_transaction_count` to become mutable too

File: zebrad/src/components/mempool/storage/eviction_list.rs (hunk 14 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: This should now be fixed, see other comment

File: zebrad/src/components/mempool/storage/eviction_list.rs (hunk 11 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: This should be fine now with `len()` fixed, see other comment

File: zebrad/src/components/mempool/storage/eviction_list.rs (hunk 12 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: Thinking about it, it seems refreshes won't ever happen in practice. Txs are only evicted from the mempool, never directly by peers (same applies to expired txs, since this `EvictedList` is being used for those too). And txs don't enter the mempool if they are in the evicted list. It's also impossible for the mempool to have two txs with the same hash (they would conflict). Therefore I think it's impossible for a refresh to happen. Does this seem right? If so, we can remove the entire "refresh" logic and the `unique`/`ordered` lists will always have the same size.

In any case, I added the limit in 8b330de

Comment (teor2345) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: I think you're right, nice catch!

Comment (conradoplg) [reply zebrad/src/components/mempool/storage/eviction_list.rs]: Simplified the logic in 3a33372

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            self.pop_front();
+        }
+        let value = Instant::now();
+        let old_value = self.unique_entries.insert(key, value);
+        // It should be impossible for an already-evicted transaction to be evicted
+        // again since transactions are not added to the mempool if they are evicted,
+        // and the mempool doesn't allow inserting two transactions with the same
+        // hash (they would conflict).
+        assert!(
+            old_value.is_none(),
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Diagnostics Nitpick: The `assert_eq!` panic message includes the actual value, but `assert!` doesn't.

```suggestion
        assert_eq!(
            old_value,
            None,
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Get the size of the list.
+    //
+    // Note: if this method being mutable becomes an issue, it's possible
+    // to compute the number of expired transactions and subtract.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Extra Docs: Let's explain performance here.

```suggestion
    // to compute the number of expired transactions and subtract,
    // at the cost of `O(len + expired)` performance each time the method is called.
    //
    // Currently the performance is `O(expired)` for the first call, then `O(1)` until the next expiry.
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+        self.ordered_entries.front()
+    }
+
+    /// Removes the first element and returns it, or `None` if the `EvictionList`
+    /// is empty.
+    fn pop_front(&mut self) -> Option<transaction::Hash> {
+        if let Some(key) = self.ordered_entries.pop_front() {
+            let removed = self.unique_entries.remove(&key);
+            assert!(
+                removed.is_some(),
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs:128]: Diagnostics Nitpick: Show the actual value.

```suggestion
            assert_eq!(
                removed,
                Some(key),
```

File: zebrad/src/components/mempool/storage/eviction_list.rs [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    /// Prune TXIDs that are older than `eviction_time` ago.
+    ///
+    // This method is public because ZIP-401 states about pruning:
+    // > This MAY be done periodically,
+    pub fn prune_old(&mut self) {
+        while let Some(txid) = self.front() {
+            let evicted_at = self
+                .unique_entries
+                .get(txid)
+                .expect("all entries should exist in both ordered_entries and unique_entries");
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage/eviction_list.rs]: Diagnostics Nitpick: Show the actual value.

Unfortunately, there isn't a nicer API for this.

```suggestion
                .unwrap_or_else(|_| panic!("all entries should exist in both ordered_entries and unique_entries, missing {:?} in unique_entries", txid));
```

**Review by teor2345 — COMMENTED** [review]
We could merge this as-is, but we're waiting on #2889.

I have a few optional suggestions for comments and diagnostics.
(The assertions should never fail, so we might never see these messages.)

**Review by teor2345 — DISMISSED** [review]
There aren't any blockers for merging this PR.

It just needs to be rebased or merged with `main` after PR #2889 merges.

**Comment by conradoplg** [comment]
> There aren't any blockers for merging this PR.
> 
> It just needs to be rebased or merged with `main` after PR #2889 merges.

Done

**Review by teor2345 — APPROVED** [review]
Looks great, let's get it merged!

File: zebrad/src/components/mempool/storage.rs [hunk 19]
```diff
@@ -142,9 +150,14 @@ impl Drop for Storage {
...
-        default.tx_cost_limit = config.tx_cost_limit;
-        default
+        Self {
+            tx_cost_limit: config.tx_cost_limit,
+            eviction_memory_time: config.eviction_memory_time,
+            verified: Default::default(),
+            tip_rejected_exact: Default::default(),
+            tip_rejected_same_effects: Default::default(),
+            chain_rejected_same_effects: Default::default(),
+        }
```
Comment (dconnolly) [inline zebrad/src/components/mempool/storage.rs:160]: 👍

File: zebrad/src/components/mempool/storage.rs [hunk 20]
```diff
@@ -103,7 +102,6 @@ pub enum RejectionError {
 }
 
 /// Hold mempool verified and rejected mempool transactions.
-#[derive(Default)]
```
Comment (dconnolly) [inline zebrad/src/components/mempool/storage.rs:106]: 👍

File: zebrad/src/components/mempool/storage.rs [hunk 21]
```diff
@@ -182,15 +195,26 @@ impl Storage {
...
-        {
+        // Once inserted, we evict transactions over the pool size limit per [ZIP-401];
+        //
+        // > On receiving a transaction: (...)
+        // > Calculate its cost. If the total cost of transactions in the mempool including this
+        // > one would `exceed mempooltxcostlimit`, then the node MUST repeatedly call
+        // > EvictTransaction (with the new transaction included as a candidate to evict) until the
+        // > total cost does not exceed `mempooltxcostlimit`.
+        //
+        // [ZIP-401]: https://zips.z.cash/zip-0401
```
Comment (dconnolly) [inline zebrad/src/components/mempool/storage.rs:206]:
```suggestion
        // Once inserted, we evict transactions over the pool size limit per [ZIP-401];
        //
        // > On receiving a transaction: (...)
        // > Calculate its cost. If the total cost of transactions in the mempool including this
        // > one would `exceed mempooltxcostlimit`, then the node MUST repeatedly call
        // > EvictTransaction (with the new transaction included as a candidate to evict) until the
        // > total cost does not exceed `mempooltxcostlimit`.
        //
        // 'EvictTransaction' is equivalent to [`VerifiedSet::evict_one()`] in our implementation.
        //
        // [ZIP-401]: https://zips.z.cash/zip-0401
```

Comment (dconnolly) [reply zebrad/src/components/mempool/storage.rs:206]: (not a blocker)

Comment (teor2345) [reply zebrad/src/components/mempool/storage.rs:206]: This is a helpful change. Would you mind opening a low-priority PR for it?
(Or cherry-picking it onto another docs PR?)

~CI is going to finish in about 5-10 minutes, and I'd like to merge this high-priority PR then.
(I have 5 high-priority PRs to merge today.)~

CI just finished and auto-merged the PR.

Comment (dconnolly) [reply zebrad/src/components/mempool/storage.rs:206]: Will do 👍

Comment (dconnolly) [reply zebrad/src/components/mempool/storage.rs:206]: Done here: https://github.com/ZcashFoundation/zebra/pull/2964
